// tradeplans/simulate.js

const object = require('utility').object;
/**
 * rtdata: { secID: xxx,price: xxx, yestClose: xxx }
 * openTrades: not closed trades
 */
module.exports = function (openTrades, rtdata, dts, endPrice, buy, sell, id) {
    var secID = rtdata.secID;
    var adjFactor = endPrice.get(secID) / rtdata.yestClose;
    var price = rtdata.price * adjFactor;
    var updatedTrades = openTrades.map((trade) => {
        var cpy = object.clone(trade);
        cpy.et = dts;
        cpy.ep = price;
        if(price > cpy.hp) {
            cpy.ht = dts;
            cpy.hp = price;
        }
        if(price < cpy.lp) {
            cpy.lt = dts;
            cpy.lp = price;
        }
        if(sell(secID, price)) {
            cpy.closed = true;
        }
        return cpy;
    })
    var newTrade = buy(secID, price) ? {
        tradeplanId: id,
        secID: secID,
        st: dts,
        ht: dts,
        lt: dts,
        et: dts,
        sp: price,
        hp: price,
        lp: price,
        ep: price,
        closed: false
    } : null;
    return {
        sell: updatedTrades,
        buy: newTrade
    };
}
