// config/index.js

const process = require('process');
const path = require('path');

/**
 * general
 */
exports.parallelN = 5;
exports.cacheCapacity = 50;
exports.sysdataDir = path.join(__dirname, '/../../../sysdata');


/**
 * http
 */
exports.access_token = 'f15e84b1cb72b925d37547a27f08682b3f74c464f93b67bd6864362c88dd47b2';
exports.isFiddler = process.argv.indexOf('-f') > -1;
exports.fiddlerHost = '127.0.0.1';
exports.fiddlerPort = 8888;

/**
 * data provider
 */
exports.bollingerK = 2;

/**
 * stock data
 */
exports.useLocalData = process.argv.indexOf('-l') > -1;
exports.stockDataDir = path.join(__dirname, '/../../../data');
exports.stockOpenTime = {
    hour: 1,
    time: 30
},
exports.stockCloseTime = {
    hour: 7,
    time: 0
}

/**
 * mongodb
 */
exports.mongoUrl = 'mongodb://localhost:27017/myproject';
exports.itemsPerPage = {
    min: 5,
    max: 20
};

/**
 * consumer
 */
exports.consumerQueryIntervals = [0, 5000, 10000, 15000, 20000];

/**
 * dispatcher
 */
exports.dispatcherHost = '127.0.0.1';
exports.dispatcherPort = 8000;
exports.checkTimeoutInterval = 5000;
exports.checkWaitingTaskInterval = 5000;
exports.maxTaskDuration = 30000;

/**
 * producer
 */
exports.produceInterval = {
    'updateStockData': 5000,
    'rtSimulate': 5000 //TODO
}
exports.produceTime = {
    'updateStockData': {
        hour: 7,
        minute: 30
    },
    'rtSimulate': {
        hour: 0,
        minute: 0
    }
}

/**
 * azure storage
 */
exports.azureUsr = {
    account: 'liuccplay',
    pwd: 'KAl0k8DrhZen3m+d+LDTY0s9DFqdDadI7sJgP2/brJ7vHHiyzgfJb7Qml3pyf3T1W9coq0bym2oZprEQYd4k5A==',
    host: 'liuccplay.blob.core.chinacloudapi.cn'
}
exports.stockdataContainer = 'stockdata';
exports.defaultContainer = 'stockdata';
exports.stockmetaContainer = 'stockmeta';
