// config/index.js

const process = require('process');
const path = require('path');

/**
 * general
 */
exports.maxCacheSize = 50;
exports.maxCacheTime = 600000;

/**
 * http
 */
exports.access_token = 'b71447597b422cbd6154e4d49feaa00f57536e5abb533909db232c29c9973c13';
exports.isFiddler = process.argv.indexOf('-f') > -1;
exports.fiddlerHost = '127.0.0.1';
exports.fiddlerPort = 8888;

/**
 * data provider
 */
exports.bollingerK = 2;
exports.macd = {
    Ns: 100,
    Nl: 200,
    Na: 80
}

/**
 * stock data
 */
exports.useLocalData = process.argv.indexOf('-l') > -1;
exports.stockDataDir = path.join(__dirname, '/../../../data');
exports.stockOpenTime = {
    hour: 1,
    time: 30
},
exports.stockCloseTime = {
    hour: 7,
    time: 0
}

/**
 * mongodb
 */
exports.mongoUrl = 'mongodb://localhost:27017/myproject';
exports.itemsPerPage = {
    min: 5,
    max: 20
};

/**
 * consumer
 */
exports.consumerQueryIntervals = [0, 5000, 10000, 15000, 20000];

/**
 * dispatcher
 */
exports.dispatcherHost = '127.0.0.1';
exports.dispatcherPort = 1337;
exports.checkTimeoutInterval = 5000;
exports.checkWaitingTaskInterval = 5000;
exports.maxTaskDuration = 30000;

/**
 * producer
 */
exports.produceInterval = {
    'updateStockData': 5000,
    'rtSimulate': 60000
}
exports.produceTime = {
    'updateStockData': {
        hour: 7,
        minute: 30
    },
    'rtSimulate': {
        hour: 0,
        minute: 0
    }
}

/**
 * azure storage
 */
exports.azureUsr = {
    account: 'yanstorage0609',
    pwd: 'HXTpjyrLqZZ2XnJ0r8sFB8JxPYg9GoXBoT3jmkUdrys4I43iR6XZM9BH1jXVgmzbD9EukAPgtSGcRGDG5Wtzvw==',
    host: 'yanstorage0609.blob.core.windows.net'
}
exports.stockdataContainer = 'stockdata';
exports.defaultContainer = 'stockdata';
exports.stockmetaContainer = 'stockmeta';
