// config/index.js

const process = require('process');
const path = require('path');

/**
 * general
 */
exports.maxCacheSize = 50;
exports.maxCacheTime = 600000;

/**
 * http
 */
exports.access_token = 'ea4e73fd12f50aa8478e0c107968679bdb19ad9880e49269d82adf5f0f25912b';
exports.isFiddler = process.argv.indexOf('-f') > -1;
exports.fiddlerHost = '127.0.0.1';
exports.fiddlerPort = 8888;

/**
 * data provider
 */
exports.bollingerK = 2;
exports.macd = {
    Ns: 100,
    Nl: 200,
    Na: 80
}

/**
 * stock data
 */
exports.useLocalData = process.argv.indexOf('-l') > -1;
exports.stockDataDir = path.join(__dirname, '/../../../data');
exports.stockOpenTime = {
    hour: 1,
    time: 30
},
exports.stockCloseTime = {
    hour: 7,
    time: 0
}

/**
 * mongodb
 */
exports.mongoUrl = 'mongodb://localhost:27017/myproject';
exports.itemsPerPage = {
    min: 5,
    max: 20
};

/**
 * consumer
 */
exports.consumerQueryIntervals = [0, 5000, 10000, 15000, 20000];

/**
 * dispatcher
 */
exports.dispatcherHost = '127.0.0.1';
exports.dispatcherPort = 1000;
exports.checkTimeoutInterval = 5000;
exports.checkWaitingTaskInterval = 5000;
exports.maxTaskDuration = 30000;

/**
 * producer
 */
exports.produceInterval = {
    'updateStockData': 5000,
    'rtSimulate': 60000
}
exports.produceTime = {
    'updateStockData': {
        hour: 7,
        minute: 30
    },
    'rtSimulate': {
        hour: 0,
        minute: 0
    }
}

/**
 * azure storage
 */
exports.azureUsr = {
    account: 'liuccplay',
    pwd: 'KAl0k8DrhZen3m+d+LDTY0s9DFqdDadI7sJgP2/brJ7vHHiyzgfJb7Qml3pyf3T1W9coq0bym2oZprEQYd4k5A==',
    host: 'liuccplay.blob.core.chinacloudapi.cn'
}
exports.stockdataContainer = 'stockdata';
exports.defaultContainer = 'stockdata';
exports.stockmetaContainer = 'stockmeta';
